Bug 1536629 - send job state and credentials from job

maleck13 opened the pull request:
**Note for Reviewer**
There are two commits here
1) making the changes required
2) adding a set of test for the provision subscriber and job. If happy with the approach here I will expand the tests to cover the other subscribers and jobs.  Making a number of changes like this, I felt a little disconcerted that I had no unit test to run so would love to add more to help protect against regressions in future changes.

These commits are independent of each other so if the new tests are not wanted I can remove.

I am currently trying this out in my local cluster but wanted to create the PR to let the integration tests run.

**Describe what this PR does and why we need it**:
The subscribers were deciding on the state of the Job, rather than the job informing the subscriber of the job state.  This pr follows on from issue https://github.com/openshift/ansible-service-broker/issues/608
Also the Job was Marshalling the extracted credentials and sending them as a string on the Msg field, then the subscriber would Unmarshal it and save it. Instead of doing this we just send the credentials through

Changes proposed in this pull request
 - simplify subscriber logic
 - send state from the individual job
 - send credentials directly via the Job rather than turning them into a JSON string and then back to credentials again in the subscriber (avoid unneeded marshalling and unmarshalling)
- Make an attempt at adding test cases for the ProvisionJob as an example for how it might be done and to get feedback on the approach 

**Which issue this PR fixes (This will close that issue when PR gets merged)**
fixes #608

maleck13 commented:
@eriknelson PR from conversation on IRC yesterday about subscribers setting state.

maleck13 commented:
This pr is also part of the larger work being done for https://github.com/openshift/ansible-service-broker/issues/475

shawn-hurley commented:
This looks really good to me, want to discuss the provisioner type being defined in the apb package.

I like the way you have made it testable by making the `provisioner` function. That looks really correct to me, and I think that we should make a not to update the `deprovision` and `update` jobs with this style.

rthallisey commented on pkg/broker/provision_subscriber.go:
```diff
@@ -17,76 +17,35 @@
 type ProvisionWorkSubscriber struct {
-	dao       *dao.Dao
-	msgBuffer <-chan JobMsg
+	dao SubcriberDAO
```
s/SubcriberDAO/SubscriberDAO/

rthallisey commented:
@mhrivnak if you change your initial commit to say ```fixes #608``` instead of ```608``` it will automatically close the git issue after this PR merges.

maleck13 commented on pkg/broker/provision_job.go:
```diff
@@ -17,64 +17,66 @@
 }
 
+// Provisioner defines a function that knows how to provision an apb
+type Provisioner func(si *apb.ServiceInstance) (string, *apb.ExtractedCredentials, error)
```
yeah no issue with that.

jmrodri commented:
I'll review this and see what the impact is on my current outstanding async work. I'm currently knee deep in this area.

mhrivnak commented:
> @mhrivnak if you change your initial commit to say fixes #608 instead of 608 it will automatically close the git issue after this PR merges.

I assume that comment is for @maleck13

maleck13 commented:
@shawn-hurley Will wait for @jmrodri review before changing the location of the provision type, in case there are further changes required. 
I think if all is good after the reviews perhaps we should merge at that point and I can follow up with the tests for the remaining subscribers and jobs? Just trying to avoid to big of a rebase and merge for others. If not I can do it on this PR not a big deal either way

shawn-hurley commented:
@maleck13 👍 for waiting for @jmrodri. If you want to follow up with the changes you can take it, but I wouldn't mind taking it off your plate if you want to work on your other PR? Completely up to you just offering my help if you want it.

maleck13 commented:
@shawn-hurley thanks for the offer. I think though I will naturally end up creating the additional tests as part of the PR work as I want to be able to run them when iterating on the feature pr and a good bit of the work for the last operation feature is around the Jobs and subscribers.

rthallisey commented:
Partially: https://github.com/openshift/ansible-service-broker/issues/542

jmrodri commented on pkg/broker/deprovision_subscriber.go:
```diff
@@ -36,64 +35,40 @@ func NewDeprovisionWorkSubscriber(dao *dao.Dao) *DeprovisionWorkSubscriber {
-				continue
+		for msg := range msgBuffer {
+			log.Debug("received deprovision message from buffer")
+			if msg.State.State == apb.StateSucceeded {
```
I would flip the check to look for the error conditions first and have escape valves, then do the success case afterwards without being in an if block.

```golang
func (d *DeprovisionWorkSubscriber) Subscribe(msgBuffer <-chan JobMsg) {
    d.msgBuffer = msgBuffer
    go func() {
        log.Info("Listening for deprovision messages")
        for msg := range msgBuffer {
            log.Debug("received deprovision message from buffer")

            if msg.State.State != apb.StateSucceeded { 
                if err := d.dao.SetState(msg.InstanceUUID, msg.State); err != nil {
                    log.Errorf("failed to set state after deprovision %v", err)
                }
                continue
            }

            instance, err := d.dao.GetServiceInstance(msg.InstanceUUID)
            if err != nil {
                log.Errorf(
                    "Error occurred getting service instance [ %s ] after deprovision job:",
                    msg.InstanceUUID,
                )   
                setFailedDeprovisionJob(d.dao, msg)
                continue
            }
            if err := cleanupDeprovision(instance, d.dao); err != nil {
                log.Errorf("Failed cleaning up deprovision after job, error: %v", err)
                // Cleanup is reporting something has gone wrong. Deprovision overall
                // has not completed. Mark the job as failed.
                setFailedDeprovisionJob(d.dao, msg)
                continue
            }   
        }
    }()
}

```

maleck13 replied:
👍

maleck13 replied:
removed the if entirely as we always want the state to be set

jmrodri commented on pkg/broker/deprovision_subscriber.go:
```diff
@@ -36,64 +35,40 @@ func NewDeprovisionWorkSubscriber(dao *dao.Dao) *DeprovisionWorkSubscriber {
-	}); err != nil {
-		log.Errorf("failed to set state after deprovision %#v", err)
+	// have to set the state here manually as the logic that triggers this is in the subscriber
+	dmsg.State.State = apb.StateFailed
```
The stutter is definitely weird. If we don't want to add the method, then `dmsg.JobState.State` might work.

maleck13 replied:
Yeah thought about doing that also, but see above comment on perhaps removing the ```JobMsg``` and just having ```JobState.State```

jmrodri commented on pkg/broker/deprovision_subscriber.go:
```diff
@@ -102,12 +77,12 @@ func cleanupDeprovision(instance *apb.ServiceInstance, dao *dao.Dao) error {
 
 	if err = dao.DeleteExtractedCredentials(id); err != nil {
-		log.Error("failed to delete extracted credentials - %#v", err)
+		log.Error("failed to delete extracted credentials - %v", err)
```
any reason you changed from `%#v` to `%v`? @shawn-hurley I think I've seen you use the `%#v` variant, thoughts?

maleck13 replied:
There seemed to be different approaches to this across the code base with ```%v``` seeming to be the most common. In my opinion, the %v is good for error messages as it will just print the error string, where as perhaps the %#v is good for debug statements as it prints the struct
```
&errors.errorString{s:"test error"}
```
see example https://play.golang.org/p/nkVaxztAG4_I

jmrodri commented on pkg/broker/provision_subscriber.go:
```diff
@@ -17,76 +17,35 @@
 )
 
-// ProvisionWorkSubscriber - Lissten for provision messages
+// ProvisionWorkSubscriber - Listen for provision messages
```
👍 to fixing typos

maleck13 commented on pkg/broker/deprovision_job.go:
```diff
@@ -43,42 +43,44 @@ func NewDeprovisionJob(serviceInstance *apb.ServiceInstance,
 func (p *DeprovisionJob) Run(token string, msgBuffer chan<- JobMsg) {
 	metrics.DeprovisionJobStarted()
+	defer metrics.DeprovisionJobFinished()
+	var jobMsg = JobMsg{
```
Nothing specific, just a habit I have as it is accessed from several blocks within the function. Happy to change it

maleck13 commented on pkg/broker/deprovision_job.go:
```diff
@@ -43,42 +43,44 @@ func NewDeprovisionJob(serviceInstance *apb.ServiceInstance,
 		log.Debug("skipping deprovision and sending complete msg to channel")
-		msgBuffer <- JobMsg{InstanceUUID: p.serviceInstance.ID.String(), PodName: "",
-			JobToken: token, SpecID: p.serviceInstance.Spec.ID, Error: ""}
+		jobMsg.State.State = apb.StateSucceeded
```
Agree on stuttering struggled with it also, but not a big fan of using setters either though. I wondered if perhaps we could remove JobMsg entirely and just have JobState as their is some duplication between them?

maleck13 replied:
On a side note, I think we could simplify things by removing the subscribers and either have a set of observers registered for each job that act on JobState changes or simply give the Job access to the DAO to set the state itself. 
Maybe better to discuss this outside of this PR though.

rthallisey replied:
I think it's worth a discussion @maleck13. I don't remember the issue where you first mentioned this idea, but it's worth bringing it up again.

maleck13 replied:
I will follow up with a clear separate issue to allow for that discussion

maleck13 commented on pkg/broker/deprovision_subscriber.go:
```diff
@@ -36,64 +35,40 @@ func NewDeprovisionWorkSubscriber(dao *dao.Dao) *DeprovisionWorkSubscriber {
-				)
-				setFailedDeprovisionJob(d.dao, msg)
-				continue
+		for msg := range msgBuffer {
```
2 reasons:
1) When you close a channel and you use the for range method, the loop also closes, where as if you just use a for loop and receive from the channel, once the channel is closed you will continue to receive the default value for that channel. If you don't want that behaviour your code would need to check the second value on a channel receive:
```
msg,closed <- chan 
if closed{
break
}
```
https://dave.cheney.net/2014/03/19/channel-axioms

2) I am ok with checking if the channel is closed, but in my opinion it also reads a little better with the range statement.

maleck13 commented:
Will rebase and bring the new Bind and Unbind jobs into line with these changes

maleck13 commented:
updated with refactored bind and unbind
added new tests for those jobs
changed the agreed on comments

@jmrodri Any further thoughts based on the comments?

I think the only outstanding things are the stutter, which I think could be resolved by removing the JobMsg and just sending through JobState (ie come up with a single type that suits the purposes). Maybe this could be a follow up issue and PR?
The other issues were the for loop ranging over the channel and the log format.

rthallisey commented:
Looks like we may need to add ```github.com/apex/log``` to glide.  Travis turned up with this error:
```
pkg/broker/binding_job.go:24:2: cannot find package "github.com/apex/log" in any of:
	/home/travis/gopath/src/github.com/openshift/ansible-service-broker/vendor/github.com/apex/log (vendor tree)
	/home/travis/.gimme/versions/go1.8.5.linux.amd64/src/github.com/apex/log (from $GOROOT)
	/home/travis/gopath/src/github.com/apex/log (from $GOPATH)
```

maleck13 commented:
I think this was an error on my part. My editor may have imported it accidentally. I have resolved that issue and rebased on master. Seems a ci job is timing out now

Craig

> On 18 Jan 2018, at 15:54, Ryan Hallisey <notifications@github.com> wrote:
> 
> Looks like we may need to add github.com/apex/log to glide. Travis turned up with this error:
> 
> pkg/broker/binding_job.go:24:2: cannot find package "github.com/apex/log" in any of:
> 	/home/travis/gopath/src/github.com/openshift/ansible-service-broker/vendor/github.com/apex/log (vendor tree)
> 	/home/travis/.gimme/versions/go1.8.5.linux.amd64/src/github.com/apex/log (from $GOROOT)
> 	/home/travis/gopath/src/github.com/apex/log (from $GOPATH)
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rthallisey commented:
Thanks Craig. I just restarted it. We'll see if it happens again.

eriknelson commented:
@maleck13 Sorry it's taken me a bit to review this. Trying to make my way through a review backlog. I'm guessing this still needs a rebase?

maleck13 commented:
rebased

maleck13 commented:
The CI build is failing, but it is not clear to me how best to find the issue. The log is quite large https://api.travis-ci.org/v3/job/330711368/log.txt

rthallisey commented:
It's failing on lint. The fix was merged an hour ago: https://github.com/openshift/ansible-service-broker/pull/662
```
=================================
              Lint               
=================================
pkg/apb/types.go:318:2: struct field HttpProxy should be HTTPProxy
pkg/apb/types.go:319:2: struct field HttpsProxy should be HTTPSProxy
Found 2 lint suggestions; failing.
Makefile:27: recipe for target 'lint' failed
make: *** [lint] Error 1
```

eriknelson commented:
Apologies. I broke the build :)

jmrodri commented:
I'll take another look at this one now that bind and unbind are in

maleck13 commented:
@rthallisey Thanks for pointer! have rebased again now

jmrodri commented:
My async bind scripts work fine with this deployed. I like the Binder interface func.
